### PR TITLE
Add support for transpile packages installed via npm

### LINKED ISSUE #706: Add support to babel transpile modules inside node_modules
Now some of us ships NPM packages (specially components) written in ES2015 without transpiling them.

That's a pretty good thing specially if they are gonna used in a project like Next.js or CRA (which does transpiling). They offer benefits like:

* No need to transpile before shipping to NPM
* Get the benefit of Webpack 2's tree-shaking

But we can't do this now we [exclude](https://github.com/zeit/next.js/blob/master/server/build/webpack.js#L166) everything inside node_modules from babel transpiling.

So, here's the proposed solution. 

We have an entry in `next.config.js` to include modules which needs to go through babel like this:

```js
module.exports = {
   babelifyModuleRoots: [
     "my-component",
     "redux/src"
   ]
}
```

### PR DESCRIPTION
**Pull request by giuseppeg**
Fixes #706 

Emits the transpiled modules in `.next/dist/node_modules` so that it works for SSR'd pages.
@arunoda / @nkzawa in case you like this solution I'll finish the PR. 
If you really want I can try to make symlinks work too.

### CODE REVIEW

**Comment by arunoda** [comment]
@giuseppeg This looks like a good approach.
I like it.

I'd like to see a test case as well.
We can simply publish a NPM package and use it.

File: server/build/webpack.js [hunk 1]
```diff
@@ -206,6 +218,10 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
     loader: 'emit-file-loader',
     include: [dir, nextPagesDir],
     exclude (str) {
+      if (shouldTranspileModule(str)) {
```
Comment (arunoda) [inline server/build/webpack.js]: This works for sure.
For the SSR version, we may need to rewrite the `require('my-module')` pointing to this emitted file.

And if that requires a ES6+ module, how we are going to handle it?

Comment (giuseppeg) [reply server/build/webpack.js]: This works for SSR since node will look for `.next/dist/node_modules` first and pick up the transpiled version in it.

> And if that requires a ES6+ module, how we are going to handle it?

During my testing `my-modules` was depending on another es6 module and both were transpiled and emitted to `.next/dist/node_modules/my-modules`. 

Later today or tomorrow (when I have time basically) I will add those to the PR together with the changes to the readme.

As I said this solution still doesn't work with linked packages though - I will try to tackle that at the end and see if it is easy/doable to fix.

Comment (arunoda) [reply server/build/webpack.js]: @giuseppeg 

> This works for SSR since node will look for .next/dist/node_modules first and pick up the transpiled version in it.

That's pretty interesting. Nice work.

Looking forward to take this in.

**Comment by rauchg** [comment]
Do we want to support some sort of signal in the module config that it definitely won't have to be transpiled?

For example, if you require a module that has a devDependency on `next`, or something like that. Or `{ engines: { next } }` or `next: { }`, or presence of `next.config.js`.

**Comment by arunoda** [comment]
@rauchg @giuseppeg transpiring nested modules is tricky. 
In that case, I'd like to see use explicit use of it.
Here's what I meant:

We look for `config.transpileModules` to get a list of modules we should transpile.
By default we don't transpile nested modules in those modules.
If the `package.json` of those modules has `transpileModules` field we use that to transpile nested modules.

Then this is clear.

**Comment by giuseppeg** [comment]
@arunoda added an example.

> By default we don't transpile nested modules in those modules.

Currently this is happening. We would need to rewrite the regexp to exclude sub dependencies like:

```
new RegExp(`${pattern.source}((?!node_modules).)*$`, pattern.flags)
```

Need to escape the original pattern probably.

> If the package.json of those modules has transpileModules field we use that to transpile nested modules.

This is tricky, I wouldn't do that but rather have people add a regexp for it.

File: server/build/webpack.js (hunk 1 above)
Comment (giuseppeg) [reply server/build/webpack.js]: @arunoda *edit* in order for this to work we need to copy the `package.json` over too. Do you have any idea how to do so? Or would you rather rewrite the imports/requires

**Comment by arunoda** [comment]
@giuseppeg 

I was trying this today. But I'm having errors.
See:

screen shot 2017-12-04 at 5 16 05 pm
screen shot 2017-12-04 at 5 16 39 pm

I think the SSR version is not using the transpiled version.

Here's how I use this:

* Run `yarn link && next build` inside a terminal in the main repo
* Go to `app` directory of the example run `yarn`
* Then run `yarn link next`
* Then start the app with `npm run dev`

**Comment by giuseppeg** [comment]
@arunoda yeah see my comment https://github.com/zeit/next.js/pull/3319#discussion_r153069977

If you copy the `package.json` of each module by hand to `.next/dist/node_modules/component-a|b` then it works.

**Comment by arunoda** [comment]
@giuseppeg ah ha.
That's won't be impossible.
I'll try.

**Comment by ScottPolhemus** [comment]
@giuseppeg @arunoda Thanks for your work on this! I've been able to work around the underlying issue in my own projects by customizing the `exclude` rules for each loader via `next.config.js`, but agree that a built-in configuration option would make the experience much better.

I took a slightly different approach in my own PR (https://github.com/zeit/next.js/pull/3470). There I've exposed a way to change the underlying `exclude` regex used by Next when creating the webpack loader configs. Curious what y'all think of that approach vs. the whitelist option used in this PR.

**Comment by giuseppeg** [comment]
@ScottPolhemus I think that your solution suffers from the same issues as the one from this PR. Is SSR working for you? Also I personally think that with your solution one would easily end up having a ton of code transpiled or emitted since it is easy to involuntarily write regexps that are too permissive i.e. whitelisting most of the times is better than blacklisting.

That said I might be missing something and be (happily) wrong.

**Comment by ScottPolhemus** [comment]
Actually @giuseppeg, unless I'm missing something I don't think the error that @arunoda ran into above is related to SSR after all.

It looks like your example page (`transpile-node-modules/app/pages/index.js`) is attempting to import `Logo` from `component-a`, but there is no default export from `component-a` for you to import. (`component-a` includes a component in a subdirectory which has `Logo` imported as a dependency from `component-b`.) If I replace the index page with the following (which I'm assuming is close to what you intended), Next is able to compile the page:

```
import Page from 'component-a/lib/zeit'

export default () => (
  <div>
    <Page />
  </div>
)
```

The actual error message was pretty misleading, but with this change your example runs fine for me.

I don't feel too strongly about using a whitelist vs. exposing the underlying `exclude` regex. I feel that this approach (the whitelist) obscures the actual webpack configuration from the user a little bit more (in the sense that it creates additional logic within Next for "magically" generating the appropriate loader rules, rather than expecting the user to actually know how webpack deals with things like the loaders' `exclude` options) which can be a good thing or a bad thing depending on your philosophy.

**Comment by giuseppeg** [comment]
Yup it works if you require the actual module. The problem though is that many people will rely on the `main` field of package.json I think and get this error. If we also emit the `package.json` we are all good :) I will finish this branch during the xmas holidays. 

> I feel that this approach (the whitelist) obscures the actual webpack configuration from the user

This is great, better expose an API so you can always change the underlying implementation. Otherwise you are locked to that solution and can't make a change without breaking stuff.

**Comment by giuseppeg** [comment]
By the way if you think that it could be useful we could discuss about an option to use the regexps for blacklisting i.e. make it configurable to support both.

File: readme.md [hunk 2]
```diff
@@ -1055,6 +1055,28 @@ module.exports = {
 
 Note: Next.js will automatically use that prefix the scripts it loads, but this has no effect whatsoever on `/static`. If you want to serve those assets over the CDN, you'll have to introduce the prefix yourself. One way of introducing a prefix that works inside your components and varies by environment is documented [in this example](https://github.com/zeit/next.js/tree/master/examples/with-universal-configuration).
 
+### Transpile NPM modules
+
+<p><details>
+  <summary><b>Examples</b></summary>
+  <ul><li><a href="./examples/with-es6-npm-module-pages">Transpile NPM modules</a></li></ul>
```
Comment (giuseppeg) [inline readme.md]: TODO: fix this link

File: server/build/webpack.js [hunk 3]
```diff
@@ -191,6 +191,36 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
...
+      const packageJsonPath = join(filePathDirname, 'package.json')
+      if (existsSync(packageJsonPath)) {
+        return packageJsonPath
+      }
+      filePathDirname = dirname(filePathDirname)
+    }
+  }
+
+  const packageJsonToCopy = {}
+  const copiedPackageJson = {}
```
Comment (giuseppeg) [inline server/build/webpack.js]: not sure if this is persisted across builds (when rebuilding)

File: server/build/loaders/emit-file-loader.js [hunk 4]
```diff
@@ -28,6 +28,13 @@ module.exports = function (content, sourceMap) {
     callback(null, code, map)
   }
 
+  if (query.copyPackageJson) {
```
Comment (giuseppeg) [inline server/build/loaders/emit-file-loader.js]: this should/could be a separate loader

**Review by giuseppeg — COMMENTED** [review]
@arunoda in my last commit you can see a way to copy the `package.json` files over. I am sure that there are edge cases where this might not work. I pushed this pof just to see what you/@rauchg think about it.

File: server/build/webpack.js [hunk 5]
```diff
@@ -203,6 +203,24 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
     })
   }
 
+  function getPackageJsonPath (filePath) {
+    let filePathDirname = dirname(filePath)
+    if (filePathDirname === dir) {
+      return join(dir, 'package.json')
```
Comment (giuseppeg) [inline server/build/webpack.js]: need to check whether this file exists or probably return `undefined` in this case.

**Comment by maxs15** [comment]
Is this PR functional @giuseppeg ? or still some work to be fully functional ?

**Comment by giuseppeg** [comment]
@maxs15 functional? yep. Done? Not sure, it needs to be reviewed, polished and test for a few edge cases. Feel free to try it out locally.

**Comment by maxs15** [comment]
Ok awesome work @giuseppeg, waiting for that feature for a long time !
I'll try it out

**Comment by maxs15** [comment]
@giuseppeg do you think there is any way of being able to transpile a file outside of node_modules ?
It looks like the exclude function is only called when the files are the include property right ?
Do you think exposing a property to customize the include property could work ?

Thanks

**Comment by maxs15** [comment]
It works but the issue that I have now is that the SSR doesn't require the compiled file in the .next directory.
Any idea which loader handle the case of requiring the file in .next if it exist ?

**Comment by jamesgorrie** [comment]
> do you think there is any way of being able to transpile a file outside of node_modules ?

This would be great if it could do that, as things like Yarn workspaces should then be easier to work.
I am going to start trying that here: https://github.com/jamesgorrie/next.js/blob/with-yarn-workspaces/examples/with-yarn-workspaces

File: server/build/webpack.js [hunk 6]
```diff
@@ -191,6 +191,36 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
...
+        const message = `Incorrect pattern in config.transpileModules: "${pattern}".` +
+          'It accepts an array of regular expression'
+        throw new Error(message)
+      }
+
+      return pattern.test(path)
+    })
+  }
+
+  function getPackageJsonPath (filePath) {
```
Comment (brettstack) [inline server/build/webpack.js]: Consider using https://www.npmjs.com/package/find-root instead

```js
function getPackageJson (filePath) {
  const pkgRoot = findRoot(filePath)
  const packageJsonPath = path.resolve(pkgRoot, 'package.json')

  return packageJsonPath
}
```

**Comment by brettstack** [comment]
I'm currently refactoring into next.js. I had this working previously by deleting the `exclude` and specifying `include` on the loader to check if the package is either inside the project (default functionality) or if the package.json specified a `module` prop. The answers on this SO provide some good links on this proposed/de-facto standard prop https://stackoverflow.com/questions/42708484/what-is-the-module-package-json-field-for if you need more context on `module`.

This is how I'm transpiling node_modules by patching in `next.config.js`:

```js
config.module.rules = config.module.rules.map(rule => {
  if (rule.use && rule.use.loader === 'babel-loader') {
    delete rule.exclude
    rule.include = (filePath) => {
      const shouldTranspile = pathIsInside(filePath, __dirname) || hasPkgModule(filePath)
      return shouldTranspile
    }
  }

  return rule
})
```

Rather, I'm trying to do this. I was doing the same thing in my webpack config previously and it was working fine. I confirmed the include rule is returning true for the module I'm trying to transpile but I'm getting `SyntaxError: Unexpected token import` (if you have any ideas, please let me know and save me some debugging time 🙏)

Maybe consider providing an option to transpile based on package.json.module also?

**Comment by giuseppeg** [comment]
@brettstack with Next 5 and universal webpack probably we need to revisit this solution. I think that many things can go wrong when one tries to transpile external dependencies so I would understand if the folks at Zeit decided to discard this idea. 

Anyway are you trying to patch Next 5's?

**Comment by timneutkens** [comment]
You can use https://github.com/zeit/next.js/pull/3732

**Comment by pozylon** [comment]
Maybe helpful, this is how meteor tackles the problem in it's newest beta:
https://github.com/meteor/meteor/pull/9559#issuecomment-376630230

**Comment by giuseppeg** [comment]
@timneutkens feel free to close this PR if you think that this solution won't work with the universal webpack rework

**Comment by timneutkens** [comment]
👍
